conformance/lang/expressions/let-expr/let_expr.balt

```diff
@@ -0,0 +1,611 @@
+                   Department[] deptList = [{id: 1, name:"HR"}] in
+                   from var person in personList
+                        join var {id: deptId, name: deptName} in deptList
+                        on person.id equals deptId
+                        select {
+                            fname : person.fname,
+                            lname : person.lname,
+                            dept : deptName
+                        };
+    io:println(l); // @output [{"fname":"Sam","lname":"George","dept":"HR"}]
+}
+
+Test-Case: error
+Description: Test let-expr when let variable is declared outside let-expr.
+Labels: int, let-expr
+
+function errorFunction() {
+    int i = 2;
+    io:println(let int i = 2 in i); // @error redeclared symbol 'i'
+}
+
+Test-Case: error
+Description: Test let-expr when the type of let variables does not compatible with let expression.
+Labels: function-call-expr, int, let-expr, optional-type, string, string-literal, var
+
+function errorFunction() {
+    io:println(let string i = "2", int j = 3 in i + j); // @error operator '+' not defined for 'string' and 'int'
+    io:println(let int i = 2 in foo(i)); // @error incompatible types: expected 'string?', found 'int'
+    io:println(let var i = 2 in foo(i)); // @error incompatible types: expected 'string?', found 'int'
+}
+
+function foo(string? s) returns string? {
+    return s;
+}
+
+Test-Case: error
+Description: Test let-expr when when the let variables are incorrect.
+Labels: int, let-expr, string-literal
+
+function errorFunction() {
+    io:println(let int i = "not int" in i); // @error incompatible types: expected 'int', found 'string'
+}
+
+Test-Case: output
+Description: Test let-expr being an isolated expression, when the expression following `in` is isolated.
+             Referenced variables bound by let-var-decls are initialized using isolated expressions.
+Labels: function-call-expr, int, let-expr
+
+isolated int x = 2;
+isolated int y = 4;
+
+isolated function bar(int i) returns int {
+    return i;
+}
+
+isolated function init() {
+    lock {
+        x = let int i = 20 in i;
+        io:println(x); // @output 20
+        x = let int i = 22 in bar(i);
+        io:println(x); // @output 22
+    }
+
+    lock {
+        int x1 = let int i = y in bar(i);
+        io:println(x1); // @output 4
+        int x2 = let int i = bar(200) in bar(i);
+        io:println(x2); // @output 200
+    }
+}
+
+Test-Case: error
+Description: Test error cases when let-expr is assigned to an isolated variable.
+             Non isolated expressions cannot be used to initialize let-var-decl when expression following in is isolated.
+Labels: array-type, function-call-expr, let-expr
+
+isolated int[] x = [];
+
+function foo() {
+    lock {
+        int[] y = [23];
+        x = let int[] i = y in i; // @error let-expr is not isolated because the expression that initialize variable i is not isolated
```

Comment:
```suggestion
        x = let int[] i = y in i; // @error the let expression is not isolated because the expression that initializes variable i is not isolated
```